build-tools/packages/build-cli/src/commands/release/prepare.ts

```diff
@@ -0,0 +1,266 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { MonoRepo, Package } from "@fluidframework/build-tools";
+import async from "async";
+import chalk from "chalk";
+import execa from "execa";
+import { ResetMode } from "simple-git";
+
+import { findPackageOrReleaseGroup, packageOrReleaseGroupArg } from "../../args.js";
+import {
+	BaseCommand,
+	type Context,
+	Repository,
+	getPreReleaseDependencies,
+} from "../../library/index.js";
+
+type CheckFunction = (
+	context: Context,
+	rgOrPkg: MonoRepo | Package,
+	print?: boolean,
+) => Promise<CheckResult>;
+
+interface CheckResult {
+	result: boolean;
+	message?: string;
+	fixCommand?: string;
+	fatal?: boolean;
+}
+
+async function checkNoLocalChanges(context: Context): Promise<CheckResult> {
+	const status = await context.gitRepo.getStatus();
+	if (status !== "") {
+		return {
+			result: false,
+			message: "There are some local changes. The branch should be clean.",
+			fixCommand: "git stash",
+			fatal: true,
+		};
+	}
+
+	return { result: true };
+}
+
+/**
+ * Returns true if all the dependencies are installed for the package/release group.
+ */
+async function checkDepsInstalled(
+	context: Context,
+	rgOrPkg: MonoRepo | Package,
+): Promise<CheckResult> {
+	const packagesToCheck = rgOrPkg instanceof MonoRepo ? rgOrPkg.packages : [rgOrPkg];
+
+	// eslint-disable-next-line @typescript-eslint/no-misused-promises -- The usage appears correct so mayeb a false positive
+	const checkDeps: async.AsyncBooleanIterator<Package> = async (
+		pkg: Package,
+	): Promise<boolean> => {
+		return pkg.checkInstall(false);
+	};
+
+	// I think the unicorn/no-array-method-this-argument lint error is a false positive. As far as I can tell, the
+	// checkDeps function is defined correctly and has the correct type. It appears to work as expected at runtime.
+	// eslint-disable-next-line unicorn/no-array-method-this-argument
+	const installed = await async.every(packagesToCheck, checkDeps);
```

Comment:
passing functions which have had a manually specified type as callbacks to other functions is fragile, since if checkDeps had optional arguments where were discarded from the type when it was given the explicit type async.AsyncBooleanIterator<Package>, and this function provides extra uncommonly used arguments (like array.every does when passing its second this augment), you can end up with incorrectly typed invalid arguments getting passed to optional parameters of the callback. There is some discussion of this in https://github.com/microsoft/FluidFramework/pull/14421/files#diff-c45b354bea51ccd8727afb1db811eb0764306f87f21982d9199cd04e633098a5R232 .

If you wrap your callback in a lambda in places like this, it thus bug can be avoided.

Note that this isn't just an issue in the current version of the code. COnider if in the future an additional optional argument were added to async.AsyncBooleanIterator or an additional parameter added to async.every: that could break this. Using wrapper lambdas in places like this makes the code more robust to future changes like that which can be very hard to track down and cause funny errors due to violating the type system.

Reply:
I don't understand how to wrap this in a lambda, or maybe I don't understand what to wrap. async.every seems to be expecting a second param that returns void. This satisfies the typing but seems wrong:

```ts
	const checkDeps = async (pkg: Package): Promise<boolean> => {
		return pkg.checkInstall(false);
	};

	const installed = await every(packagesToCheck, () => checkDeps);
```

Reply:
in `const installed = await every(packagesToCheck, () => checkDeps);` you forgot to actually invoke the function you are wrapping. The wrapper is supposed to do the same thing as the function, forwarding its arguments to it, not discard all arguments and return the function. I'm a bit surprised that type checks. Maybe it wouldn't if you actually typed `installed` as Boolean? I'm not sure what its doing.

The idea is to convert:

`referenceToSomeFunction` into `(args) => referenceToSomeFunction(args)` in contexts that could implicitly convert `referenceToSomeFunction` into a function with a different number of arguments (which is common with functions like array.every).

Applying that here would look like replacing
`const installed = await async.every(packagesToCheck, checkDeps);`
with
`const installed: boolean = await async.every(packagesToCheck, (item, callback) => checkDeps(item);`

This avoids passing the extra callback parameter to checkDeps which doesn't expect it (you can even omit that augment: doing so is safe since as extra unexpected argument to the lambda harms nothing). This protects against future changes breaking things. Imagine check deps got a second optional parameter added later (which seems like a non breaking change). The code as is would pass the callback to it, possibly breaking things.

At least I think thats what the linter is warning you about.

In this case checkDeps is strongly typed as async.AsyncBooleanIterator and private, and is itself a function not a reference to a function from elsewhere, so this is unlikely to be a problem. That said making this code more robust in a probably unnecessary way is likely cleaner than suppressing the linter and explaining why its ok in this case.

Reply:
Here are three variations, all of which seem to work at runtime in my testing, but only one doesn't require a lint disable, and it's the weirdest one IMO.

```ts
// Works at runtime and does not require a lint disable, but looks weird.
const installed: boolean = await every(packagesToCheck, () => checkDeps);

// Works at runtime but requires a lint disable.
// eslint-disable-next-line @typescript-eslint/no-misused-promises
const installed = await every(packagesToCheck, checkDeps);

// Works at runtime but requires a lint disable..
// eslint-disable-next-line @typescript-eslint/no-misused-promises
const installed: boolean = await every(packagesToCheck, async (item, callback) =>
	checkDeps(item),
);
```

The typing seems to want a void-return function as its second argument. According to [the async docs](https://caolan.github.io/async/v3/global.html#AsyncFunction):

> Wherever we accept a Node-style async function, we also directly accept an [ES2017 `async` function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function). In this case, the `async` function will not be passed a final callback argument, and any thrown error will be used as the `err` argument of the implicit callback, and the return value will be used as the `result` value. (i.e. a `rejected` of the returned Promise becomes the `err` callback argument, and a `resolved` value becomes the `result`.)

I can't put this information together in a way that makes sense. Any other suggested fixes or explanations for what a type-compatible use would even look like?

Oh and I tried upgrading types/async to see if that changed anything. It didn't - it was a patch anyway.

Reply:
I think this is whats desired by the `@typescript-eslint/no-misused-promises`. There is no reason to make a function async unless you use the await keyword. You can leave both off, or use both. In my opinion, using both here makes more sense as then you get a callstack that better matches what you would get if it was sync (nice for debugging and exception printing), and the fact that checkDeps is more clear.

```
const installed: boolean = await every(packagesToCheck, async (item, callback) =>
	await checkDeps(item),
);
```

The (in my opinion less good) alternative would be:

```
const installed: boolean = await every(packagesToCheck, (item, callback) =>
	heckDeps(item),
);
```

Reply:
Both suggestions require multiple lint disables:

```ts
	const installed: boolean = await every(
		packagesToCheck,
		// eslint-disable-next-line @typescript-eslint/no-misused-promises, no-return-await, @typescript-eslint/return-await
		async (item, callback) => await checkDeps(item),
	);

	// eslint-disable-next-line @typescript-eslint/no-misused-promises, @typescript-eslint/promise-function-async
	const installed: boolean = await every(packagesToCheck, (item, callback) => checkDeps(item));
```

Reply:
Really seems like the poor typing on the async package is making it less useful than the standard lib here.


```typescript
/**
 * Checks that the dependencies for a release group or package are installed. Returns a failure result if any
 * dependencies are missing.
 */
export const CheckDependenciesInstalled: CheckFunction = async (
	_context: Context,
	releaseGroupOrPackage: MonoRepo | Package,
): Promise<CheckResult> => {
	const packagesToCheck =
		releaseGroupOrPackage instanceof MonoRepo
			? releaseGroupOrPackage.packages
			: [releaseGroupOrPackage];

	const installChecks = await Promise.all(
		packagesToCheck.map(async (pkg) => pkg.checkInstall(false)),
	);

	// This is not a boolean coercion: its an identity function. Lint is wrong.
	// eslint-disable-next-line unicorn/prefer-native-coercion-functions
	if (installChecks.every((item) => item)) {
		return {
			message: "Some dependencies aren't installed. Try installing dependencies manually.",
			fixCommand: "pnpm install",
		};
	}
};

```